lib/secretsscanner/scan/scan.go

```diff
@@ -0,0 +1,252 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package scan
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"io"
+	"log/slog"
+	"os"
+	"path/filepath"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/crypto/ssh"
+
+	accessgraphsecretsv1pb "github.com/gravitational/teleport/api/gen/proto/go/teleport/accessgraph/v1"
+	"github.com/gravitational/teleport/api/types/accessgraph"
+)
+
+// ScannerConfig specifies parameters for the scanner.
+type ScannerConfig struct {
+	Dirs []string
+	Log  *slog.Logger
+}
+
+// NewScanner creates a new scanner.
+func NewScanner(cfg ScannerConfig) (*Scanner, error) {
+	if len(cfg.Dirs) == 0 {
+		return nil, trace.BadParameter("missing dirs")
+	}
+	if cfg.Log == nil {
+		cfg.Log = slog.Default()
+	}
+	return &Scanner{
+		dirs:        cfg.Dirs,
+		log:         cfg.Log,
+		privateKeys: make(map[string]*accessgraphsecretsv1pb.PrivateKey),
+	}, nil
+}
+
+// Scanner is a scanner that scans directories for secrets.
+type Scanner struct {
+	dirs []string
+	log  *slog.Logger
+	// privateKeys is a map of private keys found during the scan.
+	// The key is the path to the private key file and the value is the private key representation.
+	privateKeys map[string]*accessgraphsecretsv1pb.PrivateKey
+}
+
+// ScanPrivateKeys scans directories for SSH private keys.
+func (s *Scanner) ScanPrivateKeys(ctx context.Context, deviceID string) []SSHPrivateKey {
+	for _, dir := range s.dirs {
+		s.findPrivateKeys(ctx, dir, deviceID)
+	}
+
+	keys := make([]SSHPrivateKey, 0, len(s.privateKeys))
+	for path, key := range s.privateKeys {
+		keys = append(keys, SSHPrivateKey{
+			Path: path,
+			Key:  key,
+		})
+	}
+	return keys
+}
+
+// SSHPrivateKey represents an SSH private key found during the scan.
+type SSHPrivateKey struct {
+	// Path is the absolute path to the private key file.
+	Path string
+	// Key is the private key representation.
+	Key *accessgraphsecretsv1pb.PrivateKey
+}
+
+// findPrivateKeys walks through all files in a directory and its subdirectories
+// and checks if they are SSH private keys.
+func (s *Scanner) findPrivateKeys(ctx context.Context, root, deviceID string) {
+	logger := s.log.With("dir", root)
+
+	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
```

Comment:
Given the size of the Walk and the amount of I/O we'll cause, have you considered:

1. Scanning hotspots first, like /Users/{user}/.ssh, and reporting those keys right away
2. Walking with multiple goroutines
3. Conversely to 2), walking cold paths at a much lower priority to not hog the disk
4. Avoiding certain paths during walks (~/Library, node_modules, maybe even Git repositories?, etc)
5. Possibly reducing the 150 bytes "pre-read" to the bare minimum to identify a PEM file?

Also, how long does a walk take in your machine?

Reply:
That's... not as bad as I thought. Considering the size of ~/Library, .git directories, node_modules, etc, full of small files, each causing at least 150 (or 40?) bytes of I/O.

I still think it might be worth it to scan "hot spots" first, as I expect we'll hit some worst case scenarios eventually (slow machines with LOTS of files), but that could be a follow up (if you think it's worth doing).

Reply:
It's worth but not sure if it's worth now. I will revisit it once everything else lands

Reply:
Oh yes, these are all suggestions for later. Feel free to resolve.